openpmd-api fails to build

Having trouble building package `openpmd-api` on all of the following x86_64 systems with `gcc@7.3.0` and `gcc@7.4.0`:
1. CentOS 7
2. CentOS 8
3. RHEL 7
4. RHEL 8
5. Ubuntu 18.04

Using develop at commit 471f23e456e3b26a (14-March-2020)

```
$> spack install openpmd-api
...
==> Error: Failed to install openpmd-api due to ChildError: ProcessError: Command exited with status 2:
    'make' '-j8'
4 errors found in build log:
     137    /home/walker/spack/lib/spack/env/gcc/g++  -DopenPMD_EXPORTS -DopenPMD_USE_VERIFY=1 -I/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/include -I/tmp/walker/spack
            -stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/nlohmann-json-3.7.2-dt2yaxrxnaksgjto4lxcsklice
            3k3tua/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.
            4.0/zlib-1.2.11-o2viq7yriiaw6nwqpaa7ltpyzqkaonhb/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpark-variant-1.4.0-xoqx2ok47eoqhm2spapxdexund3kzzai/include -isystem /home/walk
            er/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpich-3.2.1-om2ebndbpaksaykav76dtc6xoqcyekqr/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/adios2-2.5.0-ttdhw5awwkih5nxv4
            itbwzu35m5qbkeh/include  -Wall -Wextra -Wpedantic -Wshadow -Woverloaded-virtual -Wunreachable-code -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/openPMD.dir/src/IO/HDF5/ParallelHDF5IOHandler.cpp.o -c /tmp/walker/spack-
            stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/HDF5/ParallelHDF5IOHandler.cpp
     138    [ 78%] Building CXX object CMakeFiles/openPMD.dir/src/IO/JSON/JSONIOHandler.cpp.o
     139    /home/walker/spack/lib/spack/env/gcc/g++  -DopenPMD_EXPORTS -DopenPMD_USE_VERIFY=1 -I/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/include -I/tmp/walker/spack
            -stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/nlohmann-json-3.7.2-dt2yaxrxnaksgjto4lxcsklice
            3k3tua/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.
            4.0/zlib-1.2.11-o2viq7yriiaw6nwqpaa7ltpyzqkaonhb/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpark-variant-1.4.0-xoqx2ok47eoqhm2spapxdexund3kzzai/include -isystem /home/walk
            er/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpich-3.2.1-om2ebndbpaksaykav76dtc6xoqcyekqr/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/adios2-2.5.0-ttdhw5awwkih5nxv4
            itbwzu35m5qbkeh/include  -Wall -Wextra -Wpedantic -Wshadow -Woverloaded-virtual -Wunreachable-code -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/openPMD.dir/src/IO/JSON/JSONIOHandler.cpp.o -c /tmp/walker/spack-stage/sp
            ack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/JSON/JSONIOHandler.cpp
     140    [ 81%] Building CXX object CMakeFiles/openPMD.dir/src/IO/JSON/JSONIOHandlerImpl.cpp.o
     141    /home/walker/spack/lib/spack/env/gcc/g++  -DopenPMD_EXPORTS -DopenPMD_USE_VERIFY=1 -I/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/include -I/tmp/walker/spack
            -stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/nlohmann-json-3.7.2-dt2yaxrxnaksgjto4lxcsklice
            3k3tua/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.
            4.0/zlib-1.2.11-o2viq7yriiaw6nwqpaa7ltpyzqkaonhb/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpark-variant-1.4.0-xoqx2ok47eoqhm2spapxdexund3kzzai/include -isystem /home/walk
            er/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpich-3.2.1-om2ebndbpaksaykav76dtc6xoqcyekqr/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/adios2-2.5.0-ttdhw5awwkih5nxv4
            itbwzu35m5qbkeh/include  -Wall -Wextra -Wpedantic -Wshadow -Woverloaded-virtual -Wunreachable-code -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/openPMD.dir/src/IO/JSON/JSONIOHandlerImpl.cpp.o -c /tmp/walker/spack-stag
            e/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/JSON/JSONIOHandlerImpl.cpp
     142    /tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/HDF5/HDF5IOHandler.cpp: In member function 'virtual void openPMD::HDF5IOHandlerImpl::listAttributes(openP
            MD::Writable*, openPMD::Parameter<(openPMD::Operation)17>&)':
  >> 143    /tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/HDF5/HDF5IOHandler.cpp:1537:47: error: too few arguments to function 'herr_t H5Oget_info3(hid_t, H5O_info
            2_t*, unsigned int)'
     144         status = H5Oget_info(node_id, &object_info);
     145                                                   ^
     146    In file included from /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include/H5Apublic.h:22:0,
     147                     from /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include/hdf5.h:23,
     148                     from /tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/include/openPMD/IO/HDF5/HDF5IOHandlerImpl.hpp:26,
     149                     from /tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/HDF5/HDF5IOHandler.cpp:22:

     ...

     156    /home/walker/spack/lib/spack/env/gcc/g++  -DopenPMD_EXPORTS -DopenPMD_USE_VERIFY=1 -I/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/include -I/tmp/walker/spack
            -stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/nlohmann-json-3.7.2-dt2yaxrxnaksgjto4lxcsklice
            3k3tua/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.
            4.0/zlib-1.2.11-o2viq7yriiaw6nwqpaa7ltpyzqkaonhb/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpark-variant-1.4.0-xoqx2ok47eoqhm2spapxdexund3kzzai/include -isystem /home/walk
            er/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpich-3.2.1-om2ebndbpaksaykav76dtc6xoqcyekqr/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/adios2-2.5.0-ttdhw5awwkih5nxv4
            itbwzu35m5qbkeh/include  -Wall -Wextra -Wpedantic -Wshadow -Woverloaded-virtual -Wunreachable-code -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/openPMD.dir/src/IO/ADIOS/ADIOS2IOHandler.cpp.o -c /tmp/walker/spack-stage
            /spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/ADIOS/ADIOS2IOHandler.cpp
     157    [ 89%] Building CXX object CMakeFiles/openPMD.dir/src/IO/ADIOS/ADIOS2Auxiliary.cpp.o
     158    /home/walker/spack/lib/spack/env/gcc/g++  -DopenPMD_EXPORTS -DopenPMD_USE_VERIFY=1 -I/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/include -I/tmp/walker/spack
            -stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/nlohmann-json-3.7.2-dt2yaxrxnaksgjto4lxcsklice
            3k3tua/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.
            4.0/zlib-1.2.11-o2viq7yriiaw6nwqpaa7ltpyzqkaonhb/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpark-variant-1.4.0-xoqx2ok47eoqhm2spapxdexund3kzzai/include -isystem /home/walk
            er/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpich-3.2.1-om2ebndbpaksaykav76dtc6xoqcyekqr/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/adios2-2.5.0-ttdhw5awwkih5nxv4
            itbwzu35m5qbkeh/include  -Wall -Wextra -Wpedantic -Wshadow -Woverloaded-virtual -Wunreachable-code -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/openPMD.dir/src/IO/ADIOS/ADIOS2Auxiliary.cpp.o -c /tmp/walker/spack-stage
            /spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/ADIOS/ADIOS2Auxiliary.cpp
     159    [ 91%] Building CXX object CMakeFiles/openPMD.dir/src/IO/InvalidatableFile.cpp.o
     160    /home/walker/spack/lib/spack/env/gcc/g++  -DopenPMD_EXPORTS -DopenPMD_USE_VERIFY=1 -I/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/include -I/tmp/walker/spack
            -stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/nlohmann-json-3.7.2-dt2yaxrxnaksgjto4lxcsklice
            3k3tua/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/hdf5-1.12.0-6dge5tfvm4sqeuumuqn4do7npsfkjbwa/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.
            4.0/zlib-1.2.11-o2viq7yriiaw6nwqpaa7ltpyzqkaonhb/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpark-variant-1.4.0-xoqx2ok47eoqhm2spapxdexund3kzzai/include -isystem /home/walk
            er/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/mpich-3.2.1-om2ebndbpaksaykav76dtc6xoqcyekqr/include -isystem /home/walker/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/adios2-2.5.0-ttdhw5awwkih5nxv4
            itbwzu35m5qbkeh/include  -Wall -Wextra -Wpedantic -Wshadow -Woverloaded-virtual -Wunreachable-code -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/openPMD.dir/src/IO/InvalidatableFile.cpp.o -c /tmp/walker/spack-stage/spa
            ck-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-src/src/IO/InvalidatableFile.cpp
     161    CMakeFiles/openPMD.dir/build.make:403: recipe for target 'CMakeFiles/openPMD.dir/src/IO/HDF5/HDF5IOHandler.cpp.o' failed
  >> 162    make[2]: *** [CMakeFiles/openPMD.dir/src/IO/HDF5/HDF5IOHandler.cpp.o] Error 1
     163    make[2]: *** Waiting for unfinished jobs....
     164    make[2]: Leaving directory '/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build'
     165    CMakeFiles/Makefile2:107: recipe for target 'CMakeFiles/openPMD.dir/all' failed
  >> 166    make[1]: *** [CMakeFiles/openPMD.dir/all] Error 2
     167    make[1]: Leaving directory '/tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build'
     168    Makefile:132: recipe for target 'all' failed
  >> 169    make: *** [all] Error 2

See build log for details:
  /tmp/walker/spack-stage/spack-stage-openpmd-api-0.11.0-tzwey7v2bmofo4qriul4vybqefimrmpx/spack-build-out.txt
```

Here is the full build log: 
[openpmd-api.txt](https://github.com/spack/spack/files/4335013/openpmd-api.txt)


@ax3l @tylerjereddy

--- a/var/spack/repos/builtin/packages/openpmd-api/hdf5-1.12.0.patch
+++ b/var/spack/repos/builtin/packages/openpmd-api/hdf5-1.12.0.patch
@@ -1,0 +1,23 @@
+From 61ccc18cdd478c6281466f1f77de416559234dd8 Mon Sep 17 00:00:00 2001
+From: Axel Huebl <axel.huebl@plasma.ninja>
+Date: Tue, 17 Mar 2020 10:51:20 -0700
+Subject: [PATCH] HDF5: H5Oget_info Compatibility
+
+Update to work with HDF5 1.12.0 signature.
+Macro for older releases.
+---
+ src/IO/HDF5/HDF5IOHandler.cpp | 4 ++++
+ 1 file changed, 4 insertions(+)
+
+diff --git a/src/IO/HDF5/HDF5IOHandler.cpp b/src/IO/HDF5/HDF5IOHandler.cpp
+index 7043861b..c125e1f4 100644
+--- a/src/IO/HDF5/HDF5IOHandler.cpp
++++ b/src/IO/HDF5/HDF5IOHandler.cpp
+@@ -1535,3 +1535,7 @@ void HDF5IOHandlerImpl::listAttributes(Writable* writable,
+     H5O_info_t object_info;
+     herr_t status;
++#if H5_VERSION_GE(1,12,0)
++    status = H5Oget_info(node_id, &object_info, H5O_INFO_NUM_ATTRS);
++#else
+     status = H5Oget_info(node_id, &object_info);
++#endif

--- a/var/spack/repos/builtin/packages/openpmd-api/package.py
+++ b/var/spack/repos/builtin/packages/openpmd-api/package.py
@@ -55,6 +55,10 @@ class OpenpmdApi(CMakePackage):
 
     extends('python', when='+python')
 
+    # Fix breaking HDF5 1.12.0 API
+    # https://github.com/openPMD/openPMD-api/pull/696
+    patch('hdf5-1.12.0.patch', when='@:0.11.0 +hdf5')
+
     def cmake_args(self):
         spec = self.spec
 

